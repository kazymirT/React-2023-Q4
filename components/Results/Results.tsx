import { ChildComponent } from '../ChildComponent/ChildComponents';
import styles from './styles.module.css'
import { NoResults } from '../NoResults';
import { useRouter } from 'next/router';
import { Select } from '../Select/Select';
import { Pagination } from '../Pagination/Pagination';
import { updateSearchParams } from '../utils/updateSearchParams';
import { FetchArgType, ProductResponse } from '../type/type';
import { useGetProductsByNameQuery } from '../../pages/api/getData';

export const Results = () => {
  const router = useRouter();
  const fetchArg: FetchArgType = {
    name: typeof router.query.search === 'string' ? router.query.search : ' ',
    limit: typeof router.query.limit === 'string' ? router.query.limit : '5',
    page: typeof router.query.page === 'string' ? router.query.page : '1',
  }

  const skip: number = (Number(fetchArg.page) - 1) * Number(fetchArg.limit);
  fetchArg.page = String(skip);
  
  const result = useGetProductsByNameQuery(
    fetchArg,
    {
      skip: router.isFallback,
    }
  );
  const data: ProductResponse | undefined = result.data;


  const handleSelectChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const newLimit = event.target.value;

    if (newLimit !== data?.limit) {
      setParams('limit', newLimit);
    }
  };
  const setParams = (paramName: string, paramValue: string) => {
    const newUrl = updateSearchParams(paramName, paramValue);
      router.push(newUrl);
  };
  const handlePageChange = (newPage: number, totalPages: number) => {
    if (newPage >= 1 && newPage <= totalPages) {
      setParams('page', String(newPage));
    }
  };

  return (
    <div className={styles.container} data-test="results">

      {data?.products && data.products.length > 0 && (
        <>
          <div className='control'>
            <Select limit={typeof router.query.limit === 'string' ? router.query.limit : '5'} onChange={handleSelectChange} />
            <Pagination
              limit={Number(data.limit)}
              skip={skip}
              page={Number(router.query.page)}
              total={Number(data.total)}
              onChange={handlePageChange}
            />
          </div>

          <div className={styles['items']} id="left-page">
            {data.products.map((item) => (
              <ChildComponent key={item.id} data={item} />
            ))}
          </div>
        </>
      )}

      {data?.products.length === 0 && <NoResults />}
    </div>
  );
};
