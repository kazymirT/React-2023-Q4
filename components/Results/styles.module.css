
.container {
  display: flex;
  flex-direction: column;
  width: 100%;
};


#left-page{
  margin: 0 auto;
  width: 100%;
  max-width: 1280px;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
};

.control{
  display: flex;
  justify-content: space-around;
  align-items: center;
  margin-bottom: 10px;
};

