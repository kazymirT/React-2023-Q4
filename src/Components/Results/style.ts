import styled from 'styled-components';

export const ResultsContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`;
export const PageControls = styled.div`
  display: flex;
  justify-content: space-around;
  align-items: center;
  margin-bottom: 10px;
`;

export const Items = styled.div`
  margin: 0 auto;
  width: 100%;
  max-width: 1280px;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
`;
