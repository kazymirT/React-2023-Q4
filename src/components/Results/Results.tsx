import React from "react";
import { ChildComponent } from "../ChildComponent/ChildComponents";
import styles from "./styles.module.css";
import { NoResults } from "../NoResults";
import { useRouter } from "next/router";
import { Select } from "../Select/Select";
import { Pagination } from "../Pagination/Pagination";
import { updateSearchParams } from "../utils/updateSearchParams";
import { ProductResponse } from "../type/type";

type ResultsPropsType = {
  data: ProductResponse;
};

export const Results = (props: ResultsPropsType) => {
  const router = useRouter();

  const data: ProductResponse | undefined = props.data;
  const handleSelectChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const newLimit = event.target.value;

    if (newLimit !== data?.limit) {
      setParams("limit", newLimit);
    }
  };
  const setParams = (paramName: string, paramValue: string) => {
    const newUrl = updateSearchParams(paramName, paramValue);
    router.push(newUrl);
  };
  const handlePageChange = (newPage: number, totalPages: number) => {
    if (newPage >= 1 && newPage <= totalPages) {
      setParams("page", String(newPage));
    }
  };

  return (
    <div className={styles.container} data-test="results">
      {data?.products && data.products.length > 0 && (
        <>
          <div className="control">
            <Select
              limit={
                typeof router.query.limit === "string"
                  ? router.query.limit
                  : "5"
              }
              onChange={handleSelectChange}
            />
            <Pagination
              limit={Number(data.limit)}
              page={router.query.page ? Number(router.query.page) : 1}
              total={Number(data.total)}
              onChange={handlePageChange}
            />
          </div>

          <div className={styles["items"]} id="left-page">
            {data.products.map((item) => (
              <ChildComponent key={item.id} data={item} />
            ))}
          </div>
        </>
      )}

      {data?.products.length === 0 && <NoResults />}
    </div>
  );
};
