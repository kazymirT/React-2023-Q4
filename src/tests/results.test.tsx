import React from "react";
import { describe, expect, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { createCardsListResponseMock } from "./_mocks";
import { Results } from "@/components/Results/Results";

vi.mock("next/router", () => vi.importActual("next-router-mock"));

describe("Tests for the Card List component", () => {
  it("Component renders the specified number of cards - 3", async () => {
    const searchResults = render(
      <Results data={createCardsListResponseMock(true)} />,
    );
    expect(searchResults.baseElement).toMatchSnapshot();
    expect(await screen.findAllByText(/iPhone 9/)).toHaveLength(3);
  });

  it("An appropriate message is displayed if no cards are present", async () => {
    const searchResults = render(
      <Results data={createCardsListResponseMock(false)} />,
    );

    expect(searchResults.baseElement).toMatchSnapshot();
    expect(screen.queryByText("Nothing foundddddddddddd:(")).toBeFalsy();
    expect(
      screen.getByText("The search returned no results, please try again."),
    ).toBeTruthy();
    expect(screen.queryByText(/testCard/)).toBeFalsy();
  });
});
