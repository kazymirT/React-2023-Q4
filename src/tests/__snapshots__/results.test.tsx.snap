// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Card List component > An appropriate message is displayed if no cards are present 1`] = `
<body>
  <div>
    <div
      class="_container_a19081"
      data-test="results"
    >
      <div
        class="_container_0e6f7b"
      >
        The search returned no results, please try again.
      </div>
    </div>
  </div>
</body>
`;

exports[`Tests for the Card List component > Component renders the specified number of cards - 3 1`] = `
<body>
  <div>
    <div
      class="_container_a19081"
      data-test="results"
    >
      <div
        class="control"
      >
        <div
          class="selectContainer"
        >
          <span>
            Cards per page:
          </span>
          <select>
            <option
              value="5"
            >
              5
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
            <option
              value="50"
            >
              50
            </option>
          </select>
        </div>
        <div
          class="pagination"
          data-testid="pagination"
        >
          <button
            data-testid="btn-prev"
          >
            ⇐
          </button>
          <span
            data-testid="pagination-page"
          >
            NaN
            /
            10
          </span>
          <button
            data-testid="btn-next"
          >
            ⇒
          </button>
        </div>
      </div>
      <div
        class="_items_a19081"
        id="left-page"
      >
        <a
          href="/details/1?limit=5&page=1&search="
        >
          <div
            class="_item_a67193"
          >
            <img
              alt="iPhone 9"
              class="_images_a67193"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="300"
              src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=640&q=75 2x"
              style="color: transparent;"
              width="300"
            />
            <h2
              class="_title_a67193"
            >
              iPhone 9
            </h2>
          </div>
        </a>
        <a
          href="/details/1?limit=5&page=1&search="
        >
          <div
            class="_item_a67193"
          >
            <img
              alt="iPhone 9"
              class="_images_a67193"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="300"
              src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=640&q=75 2x"
              style="color: transparent;"
              width="300"
            />
            <h2
              class="_title_a67193"
            >
              iPhone 9
            </h2>
          </div>
        </a>
        <a
          href="/details/1?limit=5&page=1&search="
        >
          <div
            class="_item_a67193"
          >
            <img
              alt="iPhone 9"
              class="_images_a67193"
              data-nimg="1"
              decoding="async"
              fetchpriority="high"
              height="300"
              src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=640&q=75 2x"
              style="color: transparent;"
              width="300"
            />
            <h2
              class="_title_a67193"
            >
              iPhone 9
            </h2>
          </div>
        </a>
      </div>
    </div>
  </div>
</body>
`;
